### PR TITLE
Add optional karatho control columns in [Data] section

### PR DESCRIPTION
**Pull request by charles-cowart**
metapool can now load and validate metagenomic sample-sheets that optionally contain karathoseq controls along w/extra columns.
A new sample-sheet class for metagenomic sample-sheets has been created and will become the new current version (101). Going forward all metagenomic sample-sheets will have the optional columns functionality.
This will be extended to metatranscriptomic and abs-quant sample-sheets once everything has been reviewed, columns have been agreed upon, etc.
TODO: Add functionality to SampleSheet.add_sample() that can dynamically add columns to the data-section and perform error-handling upon the adding of the first katharoseq sample. This is for manual creation of sample-sheet objects only.

### CODE REVIEW

**Comment by charles-cowart** [comment]
Hi everyone,

The first step is to create a new version of the metagenomic sample-sheet that can handle optional columns and still validate correctly. I created three new sample-sheets to use during testing; the new columns I took from an email Gail sent. I'd appreciate if you guys can look them over and see if any columns are missing or if the test sheets can be improved in any way. Thanks!

Once this functionality has been added, we can extend it for metatranscriptomic and abs-quant sample-sheets because the code and tests will be similar. After that we can circle back on @cbrenchy 's changes to the notebooks and metapool and make sure we preserve older tests while creating ones to handle the new functionality as well.

**Comment by charles-cowart** [comment]
The three ways to generate a SampleSheet object are to load a sample-sheet from file using load_sample_sheet(), to create an empty SampleSheet() object and populate it manually, or to use make_sample_sheet() along with a Pandas dataframe. The third method is perhaps the one most used by the wetlab. All three methods have been updated to support optional katharoseq control columns.

**Comment by charles-cowart** [comment]
@antgonza Ready for review! I'd also like a review from at least one notebook user as well, if at all possible. Thanks everyone!

File: metapool/sample_sheet.py [hunk 1]
```diff
@@ -377,73 +377,62 @@ def merge(self, sheets):
...
-        # of the following columns. If the column existed previously it will
-        # be overwritten, otherwise it will be created here. Alternate versions
-        # of the column name have already been resolved at this point.
-
-        # Note that the amplicon notebook currently generates the same values
-        # for this column. If the functionality in the notebook changes, the
-        # output will continue to be redfined with the current values here.
-        well_description = table['Project Plate'].astype(str) + "." + table[
-            'Sample'].astype(str) + "." + table['Well'].astype(str)
+    def _remap_table(self, table, strict):
```
Comment (antgonza) [inline metapool/sample_sheet.py:380]: Why remove the strict=True?

File: metapool/sample_sheet.py [hunk 2]
```diff
@@ -377,73 +377,62 @@ def merge(self, sheets):
...
+            # There may also be required columns that aren't defined in result.
 
-            out = out[subset]
+            # once all columns have been renamed to their preferred names, we
+            # must determine the proper set of column names for this sample-
+            # sheet. For legacy classes this is simply the list of columns
+            # defined in each sample-sheet version. For newer classes, this is
+            # defined at run-time and requires examining the metadata that
+            # will define the [Data] section.
+            result.to_csv('result.csv')
```
Comment (antgonza) [inline metapool/sample_sheet.py:408]: why does it need to be saved?

File: metapool/sample_sheet.py [hunk 3]
```diff
@@ -377,73 +377,62 @@ def merge(self, sheets):
...
-        # been removed and non-essential columns have been dropped.
-        out['Well_description'] = well_description
+        return result
 
-        for column in self.data_columns:
-            if column not in out.columns:
-                warnings.warn('The column %s in the sample sheet is empty' %
-                              column)
-                out[column] = ''
+    def _add_data_to_sheet(self, table, sequencer, lanes, assay, strict=True):
```
Comment (antgonza) [inline metapool/sample_sheet.py:415]: Just to confirm, this change is so it becomes part of the object, right? If yes, why does it need to be named with 
and initial `_`

File: metapool/sample_sheet.py [hunk 4]
```diff
@@ -537,6 +528,12 @@ def get_lane_number(self):
 
         return int(lanes[0])
 
+    def _get_expected_columns(self, table=None):
```
Comment (antgonza) [inline metapool/sample_sheet.py:531]: Also this one? if part of the object why `_` as prefix?

File: metapool/sample_sheet.py [hunk 5]
```diff
@@ -803,6 +816,119 @@ def __init__(self, path=None):
...
+                    'I7_Index_ID', 'index', 'I5_Index_ID', 'index2',
+                    'Sample_Project', 'Well_description']
+
+    # Names of columns needed in the sample-information-file output that is
+    # later passed into Qiita:
+    # 'katharoseq_aliquot_volume_ul',
+    # 'katharoseq_batch_information',
+    # 'katharoseq_cell_count',
+    # 'katharoseq_plate_number',
+    # 'katharoseq_strain'
```
Comment (antgonza) [inline metapool/sample_sheet.py:846]: If no need, can you delete these comments? For example: katharoseq_strain is not in any other part of this PR.

File: metapool/sample_sheet.py [hunk 6]
```diff
@@ -803,6 +816,119 @@ def __init__(self, path=None):
...
+    CARRIED_PREP_COLUMNS = ['experiment_design_description', 'i5_index_id',
+                            'i7_index_id', 'index', 'index2',
+                            'library_construction_protocol', 'sample_name',
+                            'sample_plate', 'sample_project',
+                            'well_description', 'well_id_384']
+
+    def __init__(self, path=None):
+        super().__init__(path=path)
+        self.remapper = {
+            # TODO: It seems like the optional columns need to be defined in
```
Comment (antgonza) [inline metapool/sample_sheet.py:874]: Open issue and delete? If there are ToDo's please add them as PR, not as part of the code.

File: metapool/sample_sheet.py [hunk 7]
```diff
@@ -821,18 +947,11 @@ class MetagenomicSampleSheetv100(KLSampleSheet):
...
-
-    # Note: Remove syndna_pool_number as that was part of the purpose of
-    # making this change. Also, it's always going to be empty or worse have
-    # a value that won't be checked.
     data_columns = ['Sample_ID', 'Sample_Name', 'Sample_Plate', 'well_id_384',
                     'I7_Index_ID', 'index', 'I5_Index_ID', 'index2',
                     'Sample_Project', 'Well_description']
 
-    # For now, assume only MetagenomicSampleSheetv100 (and v95, v99) contains
+    # For now, assume only MetagenomicSampleSheetv101, v100, v95, v99 contains
```
Comment (antgonza) [inline metapool/sample_sheet.py:954]: To stop adding to this list, why not write it as: `For now, assume only AbsQuantSampleSheetv10 doesn't contain 'contains_replicates', the rest do. `

File: metapool/sample_sheet.py [hunk 8]
```diff
@@ -1219,6 +1338,10 @@ def make_sample_sheet(metadata, table, sequencer, lanes, strict=True):
 
     if len(messages) == 0:
         sheet._add_metadata_to_sheet(metadata, sequencer)
+
+        # the problem is that _add_data_to_sheet() is not adding all of the
+        # required columns - is it because they're not beign supplied or it
+        # doesn't know to add them in this case? We need to find out! TODO
```
Comment (antgonza) [inline metapool/sample_sheet.py:1344]: ToDo??

File: metapool/tests/test_sample_sheet.py [hunk 9]
```diff
@@ -891,11 +894,18 @@ def test_remap_table_amplicon(self):
         message = (r'The column (I5_Index_ID|index2) '
                    r'in the sample sheet is empty')
         with self.assertWarnsRegex(UserWarning, message):
-            # because obs is generated from self.table (a pre-prep df), we
+            # because obs is generated from self.table (a pre-prep df), w
             # expect 'Well_description' to be empty since it is created and
```
Comment (antgonza) [inline metapool/tests/test_sample_sheet.py:898]: we?

File: metapool/tests/test_sample_sheet.py [hunk 10]
```diff
@@ -1731,6 +1750,440 @@ def test_metagenomic_sheet_creation(self):
...
+        # validate sheet again, this time get the list of error messages.
+        msgs = sheet.quiet_validate_and_scrub_sample_sheet()
+        msgs = [str(msg) for msg in msgs]
+
+        # assert this message is present in the results, and assume all of the
+        # other messages one would expect to see are also present.
+        self.assertIn('ErrorMessage: The TubeCode column in the Data section '
+                      'is missing', msgs)
+
+        ###
```
Comment (antgonza) [inline metapool/tests/test_sample_sheet.py:1965]: rm?

File: metapool/tests/test_sample_sheet.py [hunk 11]
```diff
@@ -1731,6 +1750,440 @@ def test_metagenomic_sheet_creation(self):
...
+        self.assertIsInstance(sheet, MetagenomicSampleSheetv101)
+        self.assertFalse(sheet.contains_katharoseq_samples())
+        obs_columns = set(sheet.samples[0].to_json().keys())
+        exp_columns = {'Sample_ID', 'Sample_Name', 'Sample_Plate',
+                       'well_id_384', 'I7_Index_ID', 'index', 'I5_Index_ID',
+                       'index2', 'Sample_Project', 'Well_description',
+                       'Lane'}
+        self.assertEqual(obs_columns, exp_columns)
+
+        ###
```
Comment (antgonza) [inline metapool/tests/test_sample_sheet.py:2108]: rm? If you need different tests, please add another tests not this none obvious separations.

File: metapool/tests/test_sample_sheet.py [hunk 12]
```diff
@@ -1731,6 +1750,440 @@ def test_metagenomic_sheet_creation(self):
...
```
Comment (antgonza) [inline metapool/tests/test_sample_sheet.py:2128]: rm?

**Review by antgonza — CHANGES_REQUESTED** [review]
a few comments.

File: metapool/tests/test_sample_sheet.py (hunk 11 above)
Comment (charles-cowart) [reply metapool/tests/test_sample_sheet.py:2108]: It's more illustrative as a single test, IMHO. However I'll break them up.

File: metapool/sample_sheet.py (hunk 8 above)
Comment (charles-cowart) [reply metapool/sample_sheet.py:1344]: Ty, good catch!

File: metapool/sample_sheet.py (hunk 3 above)
Comment (charles-cowart) [reply metapool/sample_sheet.py:415]: _remap_table() became a member function when we implemented different sample-sheet versions as child classes. However, both before and after, it's just a helper method that is only called in one location, by _add_data_to_sheet().

At different times we added a forced overwrite of well_description column to plate.sample.well_id and we added this warning to resolve other issues. It was better to migrate out those 'bonus functionalities' out of _remap_table() and into the sole caller of the method. It clarifies things with unittests as well.

File: metapool/sample_sheet.py (hunk 1 above)
Comment (charles-cowart) [reply metapool/sample_sheet.py:380]: _remap_table() is just a helper method, called in only one place by _add_data_to_sheet(self, table, sequencer, lanes, assay, strict=True). It's redundant for _remap_table() to have a default value for strict as well, since _add_data_to_sheet() already handles that.

File: metapool/sample_sheet.py (hunk 4 above)
Comment (charles-cowart) [reply metapool/sample_sheet.py:531]: _get_expected_columns() is meant to be a private method and only used by KLSampleSheet and its children. It's not something users have ever needed, and it's behavior is particular. Users should lean towards using make_sample_sheet() and load_sample_sheet() as their methods to create SampleSheet() objects and hence this method shouldn't have value to the user.

File: metapool/sample_sheet.py (hunk 6 above)
Comment (charles-cowart) [reply metapool/sample_sheet.py:874]: ty for the catch! Issue was resolved.

File: metapool/tests/test_sample_sheet.py [hunk 13]
```diff
@@ -1732,6 +1742,456 @@ def test_metagenomic_sheet_creation(self):
...
+            # added katharoseq control here.
+            ['kath0001', 'kath0001', 'sample_plate_1', 'A4', 'iTru7_107_07',
+             'CCGCCTAG', 'iTru5_01_A', 'ACCGTCAG', 'Project1_99999', 'desc']
+        ]
+
+        for row in data:
+            # For all children of Samplesheet() class, the first call to
+            # add_sample() determines the number, name, and ordering of
+            # columns in the [Data] section. Changing the columns in a
+            # subsequent call will raise an Error.
```
Comment (antgonza) [inline metapool/tests/test_sample_sheet.py:1991]: These seem like the same for-loop/comment, that lines 2075 onward; would it be easier to just make a copy of the object before the changes/test to minimize code?

**Review by antgonza — CHANGES_REQUESTED** [review]
Minor comments.

File: metapool/sample_sheet.py [hunk 14]
```diff
@@ -803,6 +815,106 @@ def __init__(self, path=None):
...
+            'Kathseq_RackID': 'Kathseq_RackID'
+        }
+
+    def contains_katharoseq_samples(self):
+        # when creating samples manually, as opposed to loading a sample-sheet
+        # from file, whether or not a sample-sheet contains katharoseq
+        # controls can change from add_sample() to add_sample() and won't be
+        # determined when MetagenomicSampleSheetv101() is created w/out a
+        # file. Hence, perform this check on demand() as opposed to once at
+        # init().
```
[the remark this answers is not in the document]
Comment (charles-cowart) [reply metapool/sample_sheet.py:883]: I would prefer katharo as well, but I based it on what I saw in @cbrenchy's PR and some legacy sample-sheets. That's the main reason I asked the others to also review the PR, but as of yet I've heard no comment either way. I think it's suitable to run with, at least for the time being. 

lower-case because Caitriona has used lower-cased 'kath' and we've used upper-case 'KATH' in the past.

File: metapool/tests/test_sample_sheet.py (hunk 13 above)
Comment (charles-cowart) [reply metapool/tests/test_sample_sheet.py:1991]: It's important to understand that the purpose of the tests is to confirm the behavior when creating a fresh, new object, first with the base set of columns and second with the base + optional columns. That being said, a new instance of KarathoseqEnabledSheetCreationTests() is being created each time a test is run so it's okay to partially initialize a SampleSheet() and make it a member of the class. I will do that now ty.

File: metapool/sample_sheet.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (cbrenchy) [reply metapool/sample_sheet.py:883]: Sorry for delayed response, I was out of office - katharo is preferred. It doesn't matter if we choose upper or lower case, but we need to stick to one for consistency.

[the remark this answers is not in the document]
Comment (antgonza) [reply metapool/sample_sheet.py:883]: Great, thank you! Just to confirm, we would need to make changes in this PR and the notebooks to support katharo vs. kath, right? If yes, @charles-cowart, do you want to do that here or in another PR?

[the remark this answers is not in the document]
Comment (cbrenchy) [reply metapool/sample_sheet.py:883]: Also, just to confirm the katharo naming scheme is as follows: "katharo".abbreviated project name.plate #.Well ID.CellCounts

From what I am seeing the naming is "kathseq.18000.0.E11", which is missing plate# and abbreviated project name. Or am I looking in the wrong place?

[the remark this answers is not in the document]
Comment (charles-cowart) [reply metapool/sample_sheet.py:883]: > Sorry for delayed response, I was out of office - katharo is preferred. It doesn't matter if we choose upper or lower case, but we need to stick to one for consistency.

No worries, thank you for your reply! You got it - katharo instead of kath.

[the remark this answers is not in the document]
Comment (charles-cowart) [reply metapool/sample_sheet.py:883]: > Great, thank you! Just to confirm, we would need to make changes in this PR and the notebooks to support katharo vs. kath, right? If yes, @charles-cowart, do you want to do that here or in another PR?

I'd like that to be a separate PR, because we will need the changes here to help do that.

[the remark this answers is not in the document]
Comment (charles-cowart) [reply metapool/sample_sheet.py:883]: @antgonza I'll issue changes for the naming scheme in the examples in a subsequent PR. I'll have to review recent work on the notebooks to confirm the changes there as well and I'll need this code to make the changes. I'd like to merge this before continuing.

File: metapool/tests/test_sample_sheet.py [hunk 15]
```diff
@@ -1863,6 +1863,10 @@ def test_katharoseq_enabled_sheet_load(self):
                'platemap_generation_date', 'project_abbreviation',
                'vol_extracted_elution_ul', 'well_id_96']
         obs = sheet2._get_expected_columns()
+
+        print("EXP: %s" % exp)
+        print("OBS: %s" % obs)
+        # CHARLIE
```
Comment (antgonza) [inline metapool/tests/test_sample_sheet.py:1869]: rm

**Review by antgonza — APPROVED** [review]
